esprima.js

```diff
@@ -2293,74 +2306,74 @@
+                } else {
+                    hasProto.value = true;
+                }
             }
 
-            value = id;
-            return node.finishProperty('init', id, value, false, true);
-        }
-        if (token.type === Token.EOF || token.type === Token.Punctuator) {
-            throwUnexpectedToken(token);
-        } else {
-            key = parseObjectPropertyKey();
             if (match(':')) {
                 lex();
                 value = parseAssignmentExpression();
-                return node.finishProperty('init', key, value, false, false);
+                return node.finishProperty('init', methodOrPropertyName[0], methodOrPropertyName[1], value, false, false);
             }
-            if (match('(')) {
-                value = parsePropertyMethodFunction();
-                return node.finishProperty('init', key, value, true, false);
+
+            if (token.type === Token.Identifier) {
+                return node.finishProperty('init',
+                    methodOrPropertyName[0],
+                    methodOrPropertyName[1],
+                    methodOrPropertyName[0],
+                    false,
+                    true);
             }
-            throwUnexpectedToken(lex());
+
+            throwUnexpectedToken(lookahead);
         }
     }
 
     function parseObjectInitialiser() {
-        var properties = [], property, name, key, kind, map = {}, toString = String, node = new Node();
+        var properties = [], hasProto = {value: false}, node = new Node();
 
         expect('{');
 
         while (!match('}')) {
-            property = parseObjectProperty();
-
-            if (property.key.type === Syntax.Identifier) {
-                name = property.key.name;
-            } else {
-                name = toString(property.key.value);
-            }
-            kind = (property.kind === 'init') ? PropertyKind.Data : (property.kind === 'get') ? PropertyKind.Get : PropertyKind.Set;
-
-            key = '$' + name;
-            if (Object.prototype.hasOwnProperty.call(map, key)) {
-                if (map[key] === PropertyKind.Data) {
-                    if (strict && kind === PropertyKind.Data) {
-                        tolerateError(Messages.StrictDuplicateProperty);
-                    } else if (kind !== PropertyKind.Data) {
-                        tolerateError(Messages.AccessorDataProperty);
-                    }
-                } else {
-                    if (kind === PropertyKind.Data) {
-                        tolerateError(Messages.AccessorDataProperty);
-                    } else if (map[key] & kind) {
-                        tolerateError(Messages.AccessorGetSet);
-                    }
-                }
-                map[key] |= kind;
-            } else {
-                map[key] = kind;
-            }
-
-            properties.push(property);
+            properties.push(parseObjectProperty(hasProto));
```

Comment:
The main difficulty right now is to provide the accurate location information especially when the tree does not carry it. Let's discuss this on Wednesday's meeting.